File: composeApp/src/androidMain/kotlin/com/dnfapps/arrmatey/ui/components/bazarr/BazarrSubtitlesSection.kt
package com.dnfapps.arrmatey.ui.components.bazarr

import android.widget.Toast
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Download
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.dnfapps.arrmatey.bazarr.api.model.BazarrSubtitle
import com.dnfapps.arrmatey.bazarr.state.BazarrMediaTarget
import com.dnfapps.arrmatey.bazarr.state.BazarrSubtitlesUiState
import com.dnfapps.arrmatey.bazarr.viewmodel.BazarrMediaSubtitlesViewModel
import com.dnfapps.arrmatey.model.OperationStatus
import com.dnfapps.arrmatey.compose.utils.breakable
import com.dnfapps.arrmatey.instances.model.InstanceType
import com.dnfapps.arrmatey.shared.MR
import com.dnfapps.arrmatey.ui.components.ContainerCard
import com.dnfapps.arrmatey.utils.koinInjectParams
import com.dnfapps.arrmatey.utils.mokoString
import dev.icerock.moko.resources.compose.painterResource

/**
 * "Subtitles" section embedded in a Sonarr episode or Radarr movie detail screen, backed by
 * the selected Bazarr instance. Renders nothing when no Bazarr instance is configured or the
 * item isn't tracked by Bazarr, so it's safe to drop into any detail screen unconditionally.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun BazarrSubtitlesSection(
    target: BazarrMediaTarget,
    modifier: Modifier = Modifier,
    viewModel: BazarrMediaSubtitlesViewModel = koinInjectParams(target)
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val operationState by viewModel.operationState.collectAsStateWithLifecycle()

    // Hide entirely when Bazarr isn't configured or isn't tracking this item.
    if (state is BazarrSubtitlesUiState.NoInstance || state is BazarrSubtitlesUiState.NotTracked) {
        return
    }

    var showSearch by remember { mutableStateOf(false) }

    val context = LocalContext.current
    var pendingSubtitle by remember { mutableStateOf<BazarrSubtitle?>(null) }
    val createDocumentLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.CreateDocument("*/*")
    ) { uri ->
        uri?.let {
            pendingSubtitle?.let { subtitle ->
                viewModel.downloadToDevice(subtitle) { bytes ->
                    if (bytes != null) {
                        context.contentResolver.openOutputStream(it)?.use { outputStream ->
                            outputStream.write(bytes)
                        }
                        Toast.makeText(context, "Subtitle downloaded", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(context, "Failed to download subtitle", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(4.dp),
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.fillMaxWidth()
        ) {
            Text(
                text = mokoString(MR.strings.bazarr_subtitles),
                style = MaterialTheme.typography.titleLarge
            )
            InstanceType.Bazarr.tabIcon?.let { icon ->
                Icon(
                    painter = painterResource(icon),
                    contentDescription = null,
                    modifier = Modifier.size(24.dp)
                )
            }
            Spacer(modifier = Modifier.weight(1f))
            Button(onClick = { showSearch = true }) {
                Icon(Icons.Default.Search, contentDescription = null, modifier = Modifier.size(18.dp))
                Text(mokoString(MR.strings.bazarr_search_subtitles))
            }
        }

        when (val s = state) {
            BazarrSubtitlesUiState.Loading -> {
                Box(modifier = Modifier.fillMaxWidth().padding(16.dp), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator(modifier = Modifier.size(28.dp))
                }
            }

            is BazarrSubtitlesUiState.Error -> {
                Text(s.message, color = MaterialTheme.colorScheme.error)
            }

            is BazarrSubtitlesUiState.Success -> {
                if (s.embedded.isNotEmpty()) {
                    EmbeddedSubtitlesCard(s.embedded)
                }
                s.present.forEach { subtitle ->
                    PresentSubtitleRow(
                        subtitle = subtitle,
                        onDownload = {
                            pendingSubtitle = subtitle
                            val fileName = subtitle.path?.substringAfterLast('/') ?: subtitle.name
                            createDocumentLauncher.launch(fileName)
                        },
                        onDelete = { viewModel.delete(subtitle) }
                    )
                }

                s.missing.forEach { language ->
                    ContainerCard(
                        modifier = Modifier.fillMaxWidth(),
                        onClick = {
                            viewModel.autoSearch(language)
                        }
                    ) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            SubtitleLanguageChip(language)
                            Text(
                                text = mokoString(MR.strings.missing),
                                color = MaterialTheme.colorScheme.error,
                                style = MaterialTheme.typography.bodyMedium
                            )
                            Spacer(Modifier.weight(1f))
                            Icon(Icons.Default.Search, null)
                        }
                    }
                }
            }

            else -> Unit
        }

        if (operationState is OperationStatus.InProgress) {
            Spacer(Modifier.height(8.dp))
            Row(verticalAlignment = Alignment.CenterVertically) {
                CircularProgressIndicator(modifier = Modifier.size(18.dp))
                Spacer(Modifier.width(8.dp))
                Text(mokoString(MR.strings.bazarr_auto_search), style = MaterialTheme.typography.bodySmall)
            }
        }
    }

    if (showSearch) {
        BazarrSubtitleSearchSheet(
            target = target,
            onDismiss = {
                showSearch = false
                viewModel.load()
            }
        )
    }
}

@Composable
private fun EmbeddedSubtitlesCard(
    embedded: List<BazarrSubtitle>
) {
    ContainerCard(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = mokoString(MR.strings.bazarr_embedded_count, embedded.count()),
            fontWeight = FontWeight.SemiBold
        )
        FlowRow(
            verticalArrangement = Arrangement.spacedBy(2.dp),
            horizontalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            embedded.forEach { subtitle ->
                SubtitleLanguageChip(
                    label = buildString {
                        append(subtitle.code2.orEmpty().uppercase().ifBlank { subtitle.name })
                        if (subtitle.forced) append(" · Forced")
                        if (subtitle.hi) append(" · HI")
                    }
                )
            }
        }
    }
}

@Composable
private fun PresentSubtitleRow(
    subtitle: BazarrSubtitle,
    onDownload: () -> Unit,
    onDelete: () -> Unit
) {
    ContainerCard {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = buildString {
                    append(subtitle.name)
                    subtitle.path?.let { path ->
                        append(" • ${path.breakable()}")
                    }
                },
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier.weight(1f)
            )
            SubtitleLanguageChip(
                label = buildString {
                    append(subtitle.code2.orEmpty().uppercase().ifBlank { subtitle.name })
                    if (subtitle.forced) append(" · Forced")
                    if (subtitle.hi) append(" · HI")
                }
            )
            Row {
                if (subtitle.isExternal) {
                    IconButton(onClick = onDownload) {
                        Icon(
                            Icons.Default.Download,
                            contentDescription = mokoString(MR.strings.bazarr_download_subtitle),
                            tint = MaterialTheme.colorScheme.primary
                        )
                    }
                    IconButton(onClick = onDelete) {
                        Icon(
                            Icons.Default.Delete,
                            contentDescription = mokoString(MR.strings.bazarr_delete_subtitle),
                            tint = MaterialTheme.colorScheme.error
                        )
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/Views/Components/BazarrSubtitlesSection.swift
//
//  BazarrSubtitlesSection.swift
//  iosApp
//

import SwiftUI
import Shared

/// "Subtitles" section embedded in the Sonarr episode and Radarr movie detail screens,
/// backed by the selected Bazarr instance. Renders nothing when Bazarr isn't configured
/// or isn't tracking the item, so it is safe to drop into any detail screen.
struct BazarrSubtitlesSection: View {
    let target: BazarrMediaTarget
    @StateObject private var viewModel: BazarrMediaSubtitlesViewModelS
    @State private var showSearch = false
    @State private var toastMessage: String? = nil

    init(target: BazarrMediaTarget) {
        self.target = target
        _viewModel = StateObject(wrappedValue: BazarrMediaSubtitlesViewModelS(target: target))
    }

    var body: some View {
        let state = viewModel.state
        if state is BazarrSubtitlesUiStateNoInstance || state is BazarrSubtitlesUiStateNotTracked {
            EmptyView()
        } else {
            VStack(alignment: .leading, spacing: 12) {
                HStack(spacing: 4) {
                    Text(MR.strings().bazarr_subtitles.localized())
                        .font(.title3)
                        .bold()
                    
                    if let icon = InstanceType.bazarr.tabIcon {
                        Image(uiImage: icon.toUIImage()!)
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .frame(width: 24, height: 24)
                    }
                    
                    Spacer()
                    
                    Button {
                        showSearch = true
                    } label: {
                        HStack(spacing: 4) {
                            Image(systemName: "magnifyingglass")
                            Text(MR.strings().bazarr_search_subtitles.localized())
                        }
                        .font(.subheadline)
                    }
                    .buttonStyle(.bordered)
                }
                .frame(maxWidth: .infinity)
                
                content(state)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .sheet(isPresented: $showSearch, onDismiss: { viewModel.load() }) {
                BazarrSubtitleSearchSheet(target: target)
            }
            .overlay(alignment: .bottom) {
                if let message = toastMessage {
                    ToastView(message: message)
                        .transition(.move(edge: .bottom).combined(with: .opacity))
                        .onAppear {
                            DispatchQueue.main.asyncAfter(deadline: .now() + 2.5) {
                                withAnimation { toastMessage = nil }
                            }
                        }
                        .padding(.bottom, 64)
                }
            }
            .animation(.easeInOut(duration: 0.3), value: toastMessage != nil)
        }
    }

    @ViewBuilder private func content(_ state: BazarrSubtitlesUiState) -> some View {
        if state is BazarrSubtitlesUiStateLoading {
            HStack {
                Spacer()
                ProgressView()
                Spacer()
            }
            .padding(16)
        } else if let error = state as? BazarrSubtitlesUiStateError {
            Text(error.message).foregroundStyle(.red)
        } else if let success = state as? BazarrSubtitlesUiStateSuccess {
            if !success.embedded.isEmpty {
                EmbeddedSubtitlesCard(embedded: success.embedded)
            }
            
            ForEach(success.present, id: \.self) { subtitle in
                PresentSubtitleRow(
                    subtitle: subtitle,
                    onDownload: {
                        viewModel.downloadToDevice(subtitle) { data in
                            if let data = data {
                                let fileName = (subtitle.path as NSString?)?.lastPathComponent ?? subtitle.name
                                shareFile(data: data, fileName: fileName)
                                toastMessage = "Subtitle downloaded"
                            } else {
                                toastMessage = "Failed to download subtitle"
                            }
                        }
                    },
                    onDelete: { viewModel.delete(subtitle) }
                )
            }

            ForEach(success.missing, id: \.self) { language in
                MissingSubtitleRow(language: language, onAutoSearch: { viewModel.autoSearch(language) })
            }

            if viewModel.operationState is OperationStatusInProgress {
                HStack(spacing: 8) {
                    ProgressView()
                    Text(MR.strings().bazarr_auto_search.localized())
                        .font(.caption)
                }
                .padding(.top, 8)
            }
        }
    }
}

private struct EmbeddedSubtitlesCard: View {
    let embedded: [BazarrSubtitle]
    
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(MR.plurals().bazarr_embedded_count.localized(embedded.count))
                .font(.subheadline)
                .fontWeight(.semibold)
            
            FlowLayout(spacing: 4) {
                ForEach(embedded, id: \.self) { subtitle in
                    SubtitleChip(label: subtitleLabel(subtitle))
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.vertical, 12)
        .padding(.horizontal, 18)
        .background(Color(.secondarySystemBackground))
        .cornerRadius(12)
    }
}

private struct PresentSubtitleRow: View {
    let subtitle: BazarrSubtitle
    let onDownload: () -> Void
    let onDelete: () -> Void
    
    var body: some View {
        HStack(spacing: 8) {
            VStack(alignment: .leading, spacing: 2) {
                Text(subtitle.name)
                    .font(.subheadline)
                if let path = subtitle.path {
                    Text(path)
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                        .lineLimit(1)
                        .truncationMode(.middle)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            
            SubtitleChip(label: subtitleLabel(subtitle))
            
            if subtitle.isExternal {
                Button(action: onDownload) {
                    Image(systemName: "square.and.arrow.down")
                        .foregroundStyle(.themePrimary)
                }
                .buttonStyle(.borderless)
                .accessibilityLabel(MR.strings().bazarr_download_subtitle.localized())

                Button(action: onDelete) {
                    Image(systemName: "trash")
                        .foregroundStyle(.red)
                }
                .buttonStyle(.borderless)
            }
        }
        .padding(.vertical, 12)
        .padding(.horizontal, 18)
        .background(Color(.secondarySystemBackground))
        .cornerRadius(12)
    }
}

private struct MissingSubtitleRow: View {
    let language: BazarrSubtitleLanguage
    let onAutoSearch: () -> Void
    
    var body: some View {
        Button(action: onAutoSearch) {
            HStack(spacing: 8) {
                SubtitleChip(label: chipLabel(language))
                Text(MR.strings().missing.localized())
                    .foregroundStyle(.red)
                    .font(.subheadline)
                
                Spacer()
                
                Image(systemName: "magnifyingglass")
                    .foregroundStyle(.secondary)
            }
        }
        .buttonStyle(.plain)
        .padding(.vertical, 12)
        .padding(.horizontal, 18)
        .background(Color(.secondarySystemBackground))
        .cornerRadius(12)
    }
}

private func subtitleLabel(_ subtitle: BazarrSubtitle) -> String {
    let code2 = subtitle.code2 ?? ""
    var label = code2.isEmpty ? subtitle.name : code2.uppercased()
    if subtitle.forced { label += " · Forced" }
    if subtitle.hi { label += " · HI" }
    return label
}
